### DISCUSSION TITLE
GPS Tutorial: fprime-gds won't launch - dumb user error?

### DISCUSSION THREAD

**Discussion by timmyd7777** [nasa/fprime#1119] [Q&A]
I'm able to build the GPS tutorial using the older repository ("gps-application" branch) here:
https://github.com/LeStarch/fprime/tree/gps-application

However, the instructions to actually run the demo fail for me.  The first step is to launch fprime-gds as follows:

`fprime-gds -d . -n`

This gives me a usage statement, and this error: `fprime-gds: error: unrecognized arguments: -d .`

The GpsApp itself does actually start, and I see console log output indicating that it successfully opened /dev/ttyACM0, but only if I launch it as superuser, i.e.

`sudo bin/Linux/GpsApp -a 127.0.0.1 -p 50000 -d /dev/ttyACM0`

I am building and running everything on a linux host computer - I'm not yet trying to run GpsApp on a separate computer from the machine running fprime-gds.

I understand that this GpsApp tutorial is several years out of date with the most current F-Prime distribution - but I am building and running everything from LeStarch's older repository (as mentioned above).  So I'd think this should work.

Is there a dumb newbie mistake that I'm making?

**Comment by LeStarch** [comment]
First off, I'd recommend looking at: https://github.com/fprime-community/gps-tutorial as this is where the new version of the code has moved.  That old branch hasn't been maintained in a long time.

To answer your specific questions: `sudo` is likely required because you don't have permissions to read `/dev/ttyACM0` on your user.....try adding your user to the `dialout` group.  This usually fixes the issue.

Next, what version of fprime-gds did you install?  fprime-gds has evolved a bit and thus those old instructions are probably out of date.  If you have the newest fprime-gds:

```
fprime-gds --dictionary </path/to/dictionary> -n
```

and the run the binary.

Since this is more of a discussion than an issue, I am going to convert this over to the Discussions.

**Comment by timmyd7777** [comment]
Thanks.  I believe I am running fprime-gds version 2.0.0.  It would help if fprime-gds had some kind of --version or -v option, or printed out its version in the --help option.  Here's what I get when typing fprime-gds --help:

```
(fprime-venv) timmyd@Mercury GpsApp % fprime-gds --help
usage: fprime-gds [-h] [-l LOGS] [--log-directly] [-g {none,html}] [--dictionary DICTIONARY] [-c CONFIG] [--gui-addr GUI_ADDR]
                  [--gui-port GUI_PORT] [--tts-port TTS_PORT] [--tts-addr TTS_ADDR] [-n] [--app APP] [-r ROOT_DIR]
                  [--ip-address ADDRESS] [--ip-port PORT] [--comm-adapter {ip}] [--comm-checksum-type {fixed,crc32}]

Run F prime deployment and GDS

optional arguments:
  -h, --help            show this help message and exit
  -l LOGS, --logs LOGS  Logging directory. Created if non-existent. [default:
                        /Users/timmyd/Projects/EarthObservant/FSW/FPrime/gps-tutorial/GpsApp/logs]
  --log-directly        Logging directory is used directly, no extra dated directories created.
  -g {none,html}, --gui {none,html}
                        Set the desired GUI system for running the deployment. [default: html]
  --dictionary DICTIONARY
                        Path to dictionary. Overrides automatic dictionary detection.
  -c CONFIG, --config CONFIG
                        Configuration for wx GUI. Ignored if not using wx.
  --gui-addr GUI_ADDR   Set the GUI server address [default: 127.0.0.1]
  --gui-port GUI_PORT   Set the GUI server address [default: 5000]
  --tts-port TTS_PORT   Set the threaded TCP socket server port [default: 50050]
  --tts-addr TTS_ADDR   set the threaded TCP socket server address [default: 0.0.0.0]
  -n, --no-app          Do not run deployment binary. Overrides --app.
  --app APP             Path to app to run. Overrides automatic app detection.
  -r ROOT_DIR, --root ROOT_DIR
                        Root directory of build artifacts, used to automatically find app and dictionary. [default: install_dest
                        field in settings.ini]
  --ip-address ADDRESS  Address of the IP adapter server. Default: 0.0.0.0
  --ip-port PORT        Port of the IP adapter server. Default: 50000
  --comm-adapter {ip}   Adapter for communicating to flight deployment. [default: ip]
  --comm-checksum-type {fixed,crc32}
                        Setup the checksum algorithm. [default: crc32]

```

I checked out and built the newer GPS tutorial from [https://github.com/fprime-community/gps-tutorial](https://github.com/fprime-community/gps-tutorial).  It all built OK, but still am not having any luck running it.  It seems as though fprime-gds won't talk to the running GpsApp.

Please note I am running them all on the same machine, not trying to talk remotely across a LAN.

Here's how I am launching GDS:
`fprime-gds --dictionary Top/GpsAppTopologyAppDictionary.xml -n --ip-address 127.0.0.1 --ip-port 50000`

Here's how I am launching the GpsApp:
`sudo build-fprime-automatic-native/bin/Linux/GpsApp -a 127.0.0.1 -p 50000 -d /dev/ttyACM0`

Sending any command from the GDS Web UI (GPS commands, NO_OPS, anything at all) fails to get a response.  In the console window where I'm running GDS, I see this output every time I attempt to send a command from the GDS Web UI:

`Uplink failed to send 20 bytes of data after 3 retries`

The console window running the GPSApp shows hundreds of lines of startup output; everything GPS-related looks normal.  The very last line of GpsApp console output is this:
`EVENT: (147) (2:1637606363,2108) WARNING_HI: (PRM) PrmFileReadError : Parameter file read failed in stage 0 with record 0 and error 1`

The GpsApp process is still clearly running.  It looks to me like GDS is just not talking to GpsApp.  And I'm probably doing something simple wrong.  Please let me know if this additional information suggests what my mistake is.  Thanks!

**Reply by timmyd7777** [reply]
OK, it all turned out to be a permissions problem after all.  I'm getting GPS telemetry all the way through to the Web UI now.

And - maintainers, you may now feel free to add Nvidia Jetson AGX to your list of supported platforms!

**Reply by timcanham** [reply]
That's great! Would you be willing to post your cmake files to [fprime-community](https://github.com/fprime-community)?

**Reply by timmyd7777** [reply]
Sure, happy to share.  Specifically, which CMake files do you want?  And where exactly should they be posted?

FWIW, I forked the "official" F-prime repo (https://github.com/nasa/fprime).  I added component subdirectories to the Ref application containing the "Math Component" demo code, and the "GPS Tutorial" demo.  I also enhanced the Math demo to add a few more scientific functions - log, exp, sin, cos, etc.

I used the Ref application as the basis for everything. I did not create a separate GpsApp for the GPS tutorial. So, my top-level Ref/CMakeLists.txt file will be specific to my version of the Ref app.  Is that still useful to you?

**Reply by timcanham** [reply]
I created a repo for your files here:

https://github.com/fprime-community/fprime-nvidia-jetson-agx

I made you a maintainer.

If you look at the other repos that are similar like:

https://github.com/fprime-community/fprime-freertos

There is a `cmake` subdirectory with `platform` and `toolchain` directories for those Cmake files. The idea is that someone who wants to build for the Jetson AGX can simply make this a sub-repo and use your cmake files.

There are parallel directories for any `Os` layer adaptations you had to make for the AGX, but if it's standard POSIX, you probably don't need them.

Finally, if you could update the README with any info about installing tools, or running containers, etc, that would be very helpful.

**Reply by timmyd7777** [reply]
Sorry for the delay here, I've been on another project for the past week (and then it was thanksgiving!)  To be clear, I never did a cross-compile build.  My builds on the Jetson AGX were always native/local builds.  The AGX is to all intents and purposes an ubuntu-18.04 linux computer, like a Raspberry Pi (but with more horsepower, and a GPU that F-Prime does not use in any way).  There is no special board support package, and I didn't have to make any tweaks to the CMake, etc. files.  'fprime-util build` worked just fine, right out of the box.

**Comment by LeStarch** [comment]
Awesome!!!!
